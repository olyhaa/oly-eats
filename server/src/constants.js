const CATEGORY = 'category';
const CUISINE = 'cuisine';
const EQUIPMENT = 'equipment';
const MEAL_TYPE = 'meal_type';

export const TAGS = {
  CATEGORY,
  CUISINE,
  EQUIPMENT,
  MEAL_TYPE,
};
