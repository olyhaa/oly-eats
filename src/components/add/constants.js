const TITLE = 'title';
const DESCRIPTION = 'description';
const PHOTO_URL = 'photo';
const SERVINGS = 'servings';

// INGREDIENTS
const INGREDIENTS = 'ingredients';
const INGREDIENTS_LABEL = 'ingredientsLabel';
const INGREDIENTS_LIST = 'ingredientsList';

// DIRECTIONS
const DIRECTIONS = 'directions';
const DIRECTIONS_LABEL = 'directionsLabel';
const DIRECTIONS_LIST = 'directionsList';

// SOURCE
const SOURCE_DISPLAY = 'sourceDisplay';
const SOURCE_URL = 'sourceURL';

// TIMING
const TIMING_PREP_VALUE_HOURS = 'timingPrepValueHours';
const TIMING_PREP_VALUE_MINS = 'timingPrepValueMins';
const TIMING_TOTAL_VALUE_HOURS = 'timingTotalValueHours';
const TIMING_TOTAL_VALUE_MINS = 'timingTotalValueMins';

// OPTIONAL FIELDS
const MEAL_TYPE = 'mealType';
const CUISINE = 'cuisine';
const EQUIPMENT = 'equipment';
const CATEGORY = 'category';

export const FIELDS = {
  TITLE,
  DESCRIPTION,
  PHOTO_URL,
  SOURCE_DISPLAY,
  SOURCE_URL,
  SERVINGS,
  TIMING_PREP_VALUE_HOURS,
  TIMING_PREP_VALUE_MINS,
  TIMING_TOTAL_VALUE_HOURS,
  TIMING_TOTAL_VALUE_MINS,
  INGREDIENTS,
  INGREDIENTS_LABEL,
  INGREDIENTS_LIST,
  DIRECTIONS,
  DIRECTIONS_LABEL,
  DIRECTIONS_LIST,
  MEAL_TYPE,
  CUISINE,
  EQUIPMENT,
  CATEGORY,
};

export const requiredFields = [
  FIELDS.TITLE,
  FIELDS.DESCRIPTION,
  FIELDS.SOURCE_DISPLAY,
  FIELDS.SERVINGS,
  FIELDS.INGREDIENTS_LIST,
  FIELDS.DIRECTIONS_LIST,
];

export const isRequired = (field) => {
  return requiredFields.includes(field);
};

export const TIMING_UNITS = {
  MINUTES: 'minutes',
  HOURS: 'hours',
};

export const isField = (field) => {
  const fieldKeys = Object.keys(FIELDS);
  for (let i = 0; i < fieldKeys.length; i++) {
    if (FIELDS[fieldKeys[i]] === field) {
      return true;
    }
  }
  return false;
};
