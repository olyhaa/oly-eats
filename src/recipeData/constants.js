const TEASPOON = 'teaspoon';
const TEASPOON_SHORT = 't';
const TABLESPOON = 'tablespoon';
const TABLESPOON_SHORT = 'T';
const CUP = 'cup';
const CUP_SHORT = 'c';
const OUNCE = 'ounce';
const OUNCE_SHORT = 'oz';
const POUND = 'pound';
const POUND_SHORT = 'lb';

const UNITS_EXACT = [
  TEASPOON_SHORT,
  TABLESPOON_SHORT,
  CUP_SHORT,
  OUNCE_SHORT,
  POUND_SHORT
];

const UNITS_MATCH = [TEASPOON, TABLESPOON, CUP, OUNCE, POUND];

export const isUnit = value => {
  let matches;
  // check if it's shorthand (must match exactly)
  matches = UNITS_EXACT.filter(unit => {
    return unit === value;
  });
  if (matches.length > 0) {
    return true;
  }

  // check if it's longhand
  matches = UNITS_MATCH.filter(unit => {
    return value.toUpperCase().includes(unit.toUpperCase());
  });
  if (matches.length > 0) {
    return true;
  }
  return false;
};
