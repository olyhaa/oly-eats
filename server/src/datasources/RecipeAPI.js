import { DataSource } from 'apollo-datasource';
import { TIMINGS } from '../constants';
import { extendResolversFromInterfaces } from 'apollo-server';

class RecipeAPI extends DataSource {
  constructor({ store }) {
    super();
    this.store = store;
  }

  async getAllRecipes() {
    const response = await this.store.Recipe.findAll();
    return Array.isArray(response)
      ? await Promise.all(
          response.map(async (recipe) => {
            const recipeObjs = await this.getRecipeData(recipe.id);
            return this.recipeReducer(recipeObjs);
          })
        )
      : [];
  }

  async getRecipe({ id }) {
    const recipeObjs = await this.getRecipeData(id);
    return this.recipeReducer(recipeObjs);
  }

  async deleteRecipe({ id }) {
    const recipe = await this.store.Recipe.findByPk(id);
    if (!recipe) {
      return this.recipeMutationReducer({
        success: false,
        message: 'ID not found',
      });
    }
    await recipe.destroy();
    return this.recipeMutationReducer({
      success: true,
    });
  }

  async addRecipe({ recipe: recipeFields }) {
    const baseRecipe = await this.store.Recipe.create(
      this.constructBaseRecipeObj(recipeFields)
    );

    const prepTimeArray = [];
    if (Array.isArray(recipeFields.prepTime)) {
      for (let i = 0; i < recipeFields.prepTime.length; i++) {
        const timeElement = recipeFields.prepTime[i];
        const timeObj = this.constructTimeObj({
          recipeid: baseRecipe.id,
          newFields: timeElement,
          type: TIMINGS.PREP_TIME,
        });
        const timeEntry = await this.store.Timing.create(timeObj);
        prepTimeArray.push(timeEntry);
      }
    }

    const totalTimeArray = [];
    if (Array.isArray(recipeFields.totalTime)) {
      for (let i = 0; i < recipeFields.totalTime.length; i++) {
        const timeElement = recipeFields.totalTime[i];
        const timeObj = this.constructTimeObj({
          recipeid: baseRecipe.id,
          newFields: timeElement,
          type: TIMINGS.TOTAL_TIME,
        });
        const timeEntry = await this.store.Timing.create(timeObj);
        totalTimeArray.push(timeEntry);
      }
    }

    const directionSections = this.addDirections({
      recipeid: baseRecipe.id,
      directions: recipeFields.directions,
    });

    /*
       const ingredients = addIngredients({
        recipeid: baseRecipe.id,
        recipeFields,
      });

      const recipeTags = await this.store.RecipeTag.create(
         constructRecipeTagObj({ recipeid: baseRecipe.id, recipeFields })
       );
*/
    return this.recipeReducer(
      {
        recipe: baseRecipe,
        prepTimeArray,
        totalTimeArray,
        diirections: directionSections,
      }
      /*
      directions,
      ingredients,
      recipeTags
      */
    );
  }

  async updateRecipe({ id, recipe: updatedFields }) {
    const recipe = await this.store.Recipe.findByPk(id);
    if (!recipe) {
      // TODO handle errors
    }
    let response = await this.store.Recipe.update(
      this.constructBaseRecipeObj(updatedFields),
      {
        where: { id: id },
      }
    );
    // TODO handle errors
    response = await this.store.Recipe.findByPk(id);
    return this.recipeReducer(response);
  }
  /*

  async addIngredients({ recipeId, recipeFields }) {
    const ingredientSection = await this.store.IngredientSection.create(
      constructIngredientSectionObj({ recipeid: baseRecipe.id, recipeFields })
    );
    const ingredient = await this.store.Ingredient.create(
      constructIngredientObj({ sectionid: ingredientSection.id, recipeFields })
    );
    const rangedAmount = await this.store.RangedAmount.create(
      constructRangedAmountObj({ ingredientid: ingredient.id, recipeFields })
    );
  }
*/

  constructBaseRecipeObj(newFields) {
    const recipeObj = {};
    if (newFields.title) {
      recipeObj.title = newFields.title;
    }
    if (newFields.description) {
      recipeObj.description = newFields.description;
    }
    if (newFields.source_display) {
      recipeObj.source_display = newFields.source_display;
    }
    if (newFields.source_url) {
      recipeObj.source_url = newFields.source_url;
    }
    if (newFields.photo_url) {
      recipeObj.photo_url = newFields.photo_url;
    }
    if (newFields.servings) {
      recipeObj.servings = newFields.servings;
    }
    return recipeObj;
  }

  constructTimeObj({ recipeid, newFields, type }) {
    const timeObj = {};
    let hasNewFields = false;

    if (newFields?.value) {
      timeObj.value = newFields.value;
      hasNewFields = true;
    }
    if (newFields?.units) {
      timeObj.units = newFields.units;
      hasNewFields = true;
    }
    if (hasNewFields) {
      timeObj.type = type;
      timeObj.recipeid = recipeid;
    }
    return timeObj;
  }

  constructDirectionSectionObj({ recipeid, directions }) {
    const directionSectionObj = {};
    if (directions.label) {
      directionSectionObj.label = directions.label;
    }
    directionSectionObj.recipeid = recipeid;
    return directionSectionObj;
  }

  constructDirectionStepObj({ sectionid, step }) {
    const stepObj = {};
    if (step.text) {
      stepObj.text = step.text;
    }
    stepObj.sectionid = sectionid;
    return stepObj;
  }

  async addDirections({ recipeid, directions }) {
    const directionSectionArray = [];

    if (Array.isArray(directions)) {
      for (let i = 0; i < directions.length; i++) {
        const section = directions[i];
        const directionSection = await this.store.DirectionSection.create(
          constructDirectionSectionObj({
            recipeid,
            section,
          })
        );

        const directionStepArray = [];
        if (Array.isArray(section.steps)) {
          for (let j = 0; i < section.steps.length; j++) {
            const step = section.steps[j];
            const directionStep = await this.store.DirectionStep.create(
              constructDirectionStepObj({
                sectionid: directionSection.id,
                step,
              })
            );
            directionStepArray.push(directionStep);
          }
        }
        directionSection.steps = directionStepArray;
        directionSectionArray.push(directionSection);
      }
    }
    return directionSectionArray;
  }

  async getRecipeData(id) {
    const recipe = await this.store.Recipe.findByPk(id);
    if (!recipe) {
      return {};
    }
    const prepTimeArray = await this.store.Timing.findAll({
      where: { recipeid: id, type: TIMINGS.PREP_TIME },
    });
    const totalTimeArray = await this.store.Timing.findAll({
      where: { recipeid: id, type: TIMINGS.TOTAL_TIME },
    });
    const directionSections = await this.store.DirectionSection.findAll({
      where: { recipeid: id },
    });

    if (directionSections) {
      for (let i = 0; i < directionSections.length; i++) {
        const directionSteps = await this.store.DirectionStep.findAll({
          where: { sectionid: directionSections[i].id },
        });
        directionSections[i].steps = directionSteps;
      }
    }

    return {
      recipe,
      prepTimeArray,
      totalTimeArray,
      directionSections,
    };
  }

  timeReducer({ timeArray }) {
    let reducedTime = [];
    if (timeArray) {
      reducedTime = timeArray.map((element) => {
        return { id: element.id, value: element.value, units: element.units };
      });
    }
    return reducedTime;
  }

  directionStepsReducer({ steps }) {
    console.log('steps array = ' + JSON.stringify(steps));
    let reducedSteps = [];
    if (steps) {
      reducedSteps = steps.map((step) => {
        return { id: step.id, text: step.text };
      });
    }
    return reducedSteps;
  }

  directionsReducer({ directionSections }) {
    console.log(
      'directionSections array = ' + JSON.stringify(directionSections)
    );
    let reducedDirections = [];
    if (directionSections) {
      reducedDirections = directionSections.map((section) => {
        const reducedSection = {};
        reducedSection.label = section.label;
        reducedSection.id = section.id;
        reducedSection.steps = this.directionStepsReducer({
          steps: section.steps,
        });
        return reducedSection;
      });
    }
    return reducedDirections;
  }

  recipeReducer({ recipe, prepTimeArray, totalTimeArray, directionSections }) {
    if (!recipe) {
      return null;
    }
    const recipeObj = {
      id: recipe.id,
      title: recipe.title,
      description: recipe.description,
      source: {
        display: recipe.source_display,
        url: recipe.source_url,
      },
      photo: recipe.photo_url,
      servings: recipe.servings,
      directions: this.directionsReducer({ directionSections }),
      timing: {
        prep: this.timeReducer({ timeArray: prepTimeArray }),
        total: this.timeReducer({ timeArray: totalTimeArray }),
      },
      dateAdded: recipe.createdAt,
      dateUpdated: recipe.updatedAt,
    };

    return recipeObj;
  }

  recipeMutationReducer({
    success = false,
    message = undefined,
    recipe = null,
  }) {
    return {
      success,
      message,
      recipe: this.recipeReducer(recipe ? { recipe } : {}),
    };
  }
}

export default RecipeAPI;
