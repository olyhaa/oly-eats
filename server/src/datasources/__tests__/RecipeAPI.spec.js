import RecipeAPI from '../RecipeAPI';
import { TIMINGS, TIME_UNITS } from '../../constants';

const mockRecipes = [
  {
    id: '1',
    title: 'my first recipe',
    description: 'the best recipe ever',
    source_display: "Mandy's Kitchen",
    source_url: 'http://some.url.com/recipe',
    photo_url: 'http://some.url.com/pic.jpg',
    servings: 2,
    createdAt: '2020-05-10 00:00:45.511 +00:00',
    updatedAt: '2020-05-10 00:00:45.511 +00:00',
  },
];

const mockStore = {
  Recipe: {
    create: jest.fn(),
    findAll: jest.fn(),
    findByPk: jest.fn(),
  },
};
const recipeDatasource = new RecipeAPI({ store: mockStore });

describe.skip('recipe reducer', () => {});

describe.skip('recipe mutation reducer', () => {});

describe('getAllRecipes', () => {
  it('returns empty array for empty db', async () => {
    mockStore.Recipe.findAll.mockReturnValueOnce([]);

    const response = await recipeDatasource.getAllRecipes();

    expect(mockStore.Recipe.findAll).toBeCalledWith();
    expect(response).toEqual([]);
  });

  it('returns array of recipes', async () => {
    mockStore.Recipe.findAll.mockReturnValueOnce(mockRecipes);

    const response = await recipeDatasource.getAllRecipes();

    expect(mockStore.Recipe.findAll).toBeCalledWith();
    expect(response).toMatchSnapshot();
  });
});

describe('getRecipe', () => {
  it('returns null when id does not exist', async () => {
    mockStore.Recipe.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad_id';

    const response = await recipeDatasource.getRecipe({ id: bad_id });

    expect(mockStore.Recipe.findByPk).toBeCalledWith(bad_id);
    expect(response).toEqual(null);
  });

  it('returns recipe', async () => {
    mockStore.Recipe.findByPk.mockReturnValueOnce(mockRecipes[0]);

    const response = await recipeDatasource.getRecipe({
      id: mockRecipes[0].id,
    });

    expect(mockStore.Recipe.findByPk).toBeCalledWith(mockRecipes[0].id);
    expect(response).toMatchSnapshot();
  });
});

describe('deleteRecipe', () => {
  it('calls store destroy and returns result - id exists', async () => {
    const findPkVal = Object.assign({}, mockRecipes[0]);
    findPkVal.destroy = jest.fn();

    mockStore.Recipe.findByPk.mockReturnValueOnce(findPkVal);
    findPkVal.destroy.mockReturnValueOnce('done');

    // check the result of the fn
    const res = await recipeDatasource.deleteRecipe({ id: findPkVal.id });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.Recipe.findByPk).toBeCalledWith(findPkVal.id);
    expect(findPkVal.destroy).toBeCalled();
  });

  it('calls store destroy and returns result - id not found', async () => {
    mockStore.Recipe.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad id';

    // check the result of the fn
    const res = await recipeDatasource.deleteRecipe({ id: bad_id });
    expect(res).toMatchSnapshot();

    expect(mockStore.Recipe.findByPk).toBeCalledWith(bad_id);
  });
});

describe.skip('addRecipe', () => {});
describe.skip('updateRecipe', () => {});

describe('constructBaseRecipeObj', () => {
  describe('no base recipe fields', () => {
    it('empty object', () => {
      const newFields = {};
      expect(recipeDatasource.constructBaseRecipeObj(newFields)).toEqual({});
    });
    it('object without relevant fields', () => {
      const newFields = { some_garbage_field: 'with some garbage values' };
      expect(recipeDatasource.constructBaseRecipeObj(newFields)).toEqual({});
    });
  });

  describe('only base recipe fields', () => {
    it('has all fields', () => {
      const newFields = {
        title: 'my title',
        description: 'some description',
        source_display: 'My Recipe source',
        source_url: 'http://source.com',
        photo_url: 'http://photos.com/mine.jpg',
        servings: 4,
      };
      expect(recipeDatasource.constructBaseRecipeObj(newFields)).toEqual(
        newFields
      );
    });

    it('has subset of fields', () => {
      const newFields = {
        title: 'my title',
        description: 'some description',
      };
      expect(recipeDatasource.constructBaseRecipeObj(newFields)).toEqual(
        newFields
      );
    });
  });

  it('base recipe fields + others', () => {
    const baseFields = {
      title: 'my title',
      description: 'some description',
      source_display: 'My Recipe source',
      source_url: 'http://source.com',
      photo_url: 'http://photos.com/mine.jpg',
      servings: 4,
    };
    const allFields = Object.assign({}, baseFields);
    allFields.some_garbage_field = 'some garbage';
    expect(recipeDatasource.constructBaseRecipeObj(allFields)).toEqual(
      baseFields
    );
  });
});

describe('constructTimeObj', () => {
  describe('no time fields', () => {
    it('empty object', () => {
      const newFields = {};
      expect(
        recipeDatasource.constructTimeObj({
          recipeid: '1',
          newFields,
          type: TIMINGS.PREP,
        })
      ).toEqual({});
    });

    it('undefined object', () => {
      const newFields = undefined;
      expect(
        recipeDatasource.constructTimeObj({
          recipeid: '1',
          newFields,
          type: TIMINGS.PREP,
        })
      ).toEqual({});
    });

    it('object without relevant fields', () => {
      const newFields = { some_garbage_field: 'with some garbage values' };
      expect(
        recipeDatasource.constructTimeObj({
          recipeid: '1',
          newFields,
          type: TIMINGS.PREP,
        })
      ).toEqual({});
    });
  });

  describe('only time fields', () => {
    it('has all fields', () => {
      const newFields = {
        value: '2',
        units: TIME_UNITS.MINUTE,
      };
      const expectedFields = Object.assign({}, newFields);
      expectedFields.type = TIMINGS.PREP;
      expectedFields.recipeid = '1';

      expect(
        recipeDatasource.constructTimeObj({
          recipeid: expectedFields.recipeid,
          newFields,
          type: expectedFields.type,
        })
      ).toEqual(expectedFields);
    });

    it('has subset of fields', () => {
      const newFields = {
        units: TIME_UNITS.MINUTE,
      };
      const expectedFields = Object.assign({}, newFields);
      expectedFields.type = TIMINGS.PREP;
      expectedFields.recipeid = '1';
      expect(
        recipeDatasource.constructTimeObj({
          recipeid: expectedFields.recipeid,
          newFields,
          type: expectedFields.type,
        })
      ).toEqual(expectedFields);
    });
  });

  it('time fields + others', () => {
    const baseFields = {
      value: '2',
      units: TIME_UNITS.MINUTE,
    };
    const allFields = Object.assign({}, baseFields);
    allFields.some_garbage_field = 'some garbage';
    const expectedFields = Object.assign({}, baseFields);
    expectedFields.type = TIMINGS.PREP;
    expectedFields.recipeid = '1';
    expect(
      recipeDatasource.constructTimeObj({
        recipeid: expectedFields.recipeid,
        newFields: allFields,
        type: expectedFields.type,
      })
    ).toEqual(expectedFields);
  });
});
