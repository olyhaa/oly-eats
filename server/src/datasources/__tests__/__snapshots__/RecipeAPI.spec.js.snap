// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addRecipe calls store creator and returns result 1`] = `
Object {
  "message": undefined,
  "recipe": Object {
    "dateAdded": "2020-05-10 00:00:45.511 +00:00",
    "dateUpdated": "2020-05-10 00:01:45.511 +00:00",
    "description": "the best recipe ever",
    "directions": Array [
      Object {
        "label": "section 1",
        "steps": Array [
          Object {
            "text": "step 1",
          },
          Object {
            "text": "step 2",
          },
        ],
      },
    ],
    "id": 123,
    "ingredients": Array [
      Object {
        "ingredients": Array [
          Object {
            "amount": "2",
            "name": "apples",
            "optional": true,
            "prep": "chopped",
            "toTaste": true,
            "unit": "cup",
          },
          Object {
            "name": "garlic",
            "optional": false,
            "prep": "minced",
            "rangedAmount": Object {
              "max": "10",
              "min": "5",
            },
            "toTaste": true,
            "unit": "tablespoon",
          },
        ],
        "label": "ingredient section 1",
      },
    ],
    "photo": "http://some.url.com/pic.jpg",
    "servings": 2,
    "source": Object {
      "display": "Mandy's Kitchen",
      "url": "http://some.url.com/recipe",
    },
    "tags": Array [
      Object {
        "id": 1,
        "label": "tag 1",
        "typeid": 1,
      },
      Object {
        "id": 2,
        "label": "tag 2",
        "typeid": 1,
      },
    ],
    "timing": Object {
      "prep": Array [
        Object {
          "units": "MINUTE",
          "value": "2",
        },
      ],
      "total": Array [
        Object {
          "units": "MINUTE",
          "value": "20",
        },
        Object {
          "units": "HOUR",
          "value": "1",
        },
      ],
    },
    "title": "my first recipe",
  },
  "success": true,
}
`;

exports[`addTags multiple tags 1`] = `
Array [
  Object {
    "createdAt": "2020-05-10 00:07:45.511 +00:00",
    "id": 1,
    "recipeId": 123,
    "tagid": 1,
    "updatedAt": "2020-05-10 00:08:45.511 +00:00",
  },
  Object {
    "createdAt": "2020-05-10 00:07:45.511 +00:00",
    "id": 2,
    "recipeId": 123,
    "tagid": 23,
    "updatedAt": "2020-05-10 00:08:45.511 +00:00",
  },
]
`;

exports[`constructDirections multiple direction sections 1`] = `
Array [
  Object {
    "directionSteps": Array [
      Object {
        "text": "step 1",
      },
    ],
    "label": "section 1",
  },
  Object {
    "directionSteps": Array [
      Object {
        "text": "step 1",
      },
      Object {
        "text": "step 2",
      },
    ],
  },
]
`;

exports[`constructDirections single direction section, all fields 1`] = `
Array [
  Object {
    "directionSteps": Array [
      Object {
        "text": "step 1",
      },
    ],
    "label": "section 1",
  },
]
`;

exports[`constructDirections single section, no steps provided, only label 1`] = `
Array [
  Object {
    "directionSteps": Array [],
    "label": "section 1",
  },
]
`;

exports[`constructDirections single section, only steps provided, no label 1`] = `
Array [
  Object {
    "directionSteps": Array [
      Object {
        "text": "step 1",
      },
      Object {
        "text": "step 2",
      },
    ],
  },
]
`;

exports[`constructIngredientObj only ingredient fields has subset of fields 1`] = `
Object {
  "name": "apples",
  "prep": "chopped",
}
`;

exports[`constructIngredientObj only ingredient fields has subset of fields that are false 1`] = `
Object {
  "name": "apples",
  "optional": false,
  "prep": "chopped",
  "toTaste": false,
}
`;

exports[`constructIngredients multiple ingredient sections 1`] = `
Array [
  Object {
    "ingredients": Array [
      Object {
        "amount": "2",
        "name": "apples",
        "optional": true,
        "prep": "chopped",
        "toTaste": true,
        "unit": "cup",
      },
    ],
    "label": "ingredient section 1",
  },
  Object {
    "ingredients": Array [
      Object {
        "amount": "1",
        "name": "garlic",
        "prep": "minced",
        "unit": "tablespoon",
      },
    ],
  },
]
`;

exports[`constructIngredients single ingredient section, all fields 1`] = `
Array [
  Object {
    "ingredients": Array [
      Object {
        "amount": "2",
        "name": "apples",
        "prep": "chopped",
        "unit": "cup",
      },
    ],
    "label": "ingredient section 1",
  },
]
`;

exports[`constructIngredients single section, no ingredients provided, only label 1`] = `
Array [
  Object {
    "ingredients": Array [],
    "label": "ingredient section 1",
  },
]
`;

exports[`constructIngredients single section, only ingredients provided, no label 1`] = `
Array [
  Object {
    "ingredients": Array [
      Object {
        "amount": "2",
        "name": "apples",
        "optional": true,
        "prep": "chopped",
        "toTaste": true,
        "unit": "cup",
      },
      Object {
        "name": "garlic",
        "optional": false,
        "prep": "minced",
        "rangedAmount": Object {
          "max": "10",
          "min": "5",
        },
        "toTaste": true,
        "unit": "tablespoon",
      },
    ],
  },
]
`;

exports[`deleteRecipe calls store destroy and returns result - id exists 1`] = `
Object {
  "message": undefined,
  "recipe": null,
  "success": true,
}
`;

exports[`deleteRecipe calls store destroy and returns result - id not found 1`] = `
Object {
  "message": "ID not found",
  "recipe": null,
  "success": false,
}
`;

exports[`getAllRecipes returns array of recipes 1`] = `
Array [
  Object {
    "dateAdded": "2020-05-10 00:00:45.511 +00:00",
    "dateUpdated": "2020-05-10 00:01:45.511 +00:00",
    "description": "the best recipe ever",
    "directions": Array [
      Object {
        "label": "section 1",
        "steps": Array [
          Object {
            "text": "step 1",
          },
          Object {
            "text": "step 2",
          },
        ],
      },
    ],
    "id": 123,
    "ingredients": Array [
      Object {
        "ingredients": Array [
          Object {
            "amount": "2",
            "name": "apples",
            "optional": true,
            "prep": "chopped",
            "toTaste": true,
            "unit": "cup",
          },
          Object {
            "name": "garlic",
            "optional": false,
            "prep": "minced",
            "rangedAmount": Object {
              "max": "10",
              "min": "5",
            },
            "toTaste": true,
            "unit": "tablespoon",
          },
        ],
        "label": "ingredient section 1",
      },
    ],
    "photo": "http://some.url.com/pic.jpg",
    "servings": 2,
    "source": Object {
      "display": "Mandy's Kitchen",
      "url": "http://some.url.com/recipe",
    },
    "tags": Array [
      Object {
        "id": 1,
        "label": "tag 1",
        "typeid": 1,
      },
      Object {
        "id": 2,
        "label": "tag 2",
        "typeid": 1,
      },
    ],
    "timing": Object {
      "prep": Array [
        Object {
          "units": "MINUTE",
          "value": "2",
        },
      ],
      "total": Array [
        Object {
          "units": "MINUTE",
          "value": "20",
        },
        Object {
          "units": "HOUR",
          "value": "1",
        },
      ],
    },
    "title": "my first recipe",
  },
]
`;

exports[`getRecipe returns recipe 1`] = `
Object {
  "dateAdded": "2020-05-10 00:00:45.511 +00:00",
  "dateUpdated": "2020-05-10 00:01:45.511 +00:00",
  "description": "the best recipe ever",
  "directions": Array [
    Object {
      "label": "section 1",
      "steps": Array [
        Object {
          "text": "step 1",
        },
        Object {
          "text": "step 2",
        },
      ],
    },
  ],
  "id": 123,
  "ingredients": Array [
    Object {
      "ingredients": Array [
        Object {
          "amount": "2",
          "name": "apples",
          "optional": true,
          "prep": "chopped",
          "toTaste": true,
          "unit": "cup",
        },
        Object {
          "name": "garlic",
          "optional": false,
          "prep": "minced",
          "rangedAmount": Object {
            "max": "10",
            "min": "5",
          },
          "toTaste": true,
          "unit": "tablespoon",
        },
      ],
      "label": "ingredient section 1",
    },
  ],
  "photo": "http://some.url.com/pic.jpg",
  "servings": 2,
  "source": Object {
    "display": "Mandy's Kitchen",
    "url": "http://some.url.com/recipe",
  },
  "tags": Array [
    Object {
      "id": 1,
      "label": "tag 1",
      "typeid": 1,
    },
    Object {
      "id": 2,
      "label": "tag 2",
      "typeid": 1,
    },
  ],
  "timing": Object {
    "prep": Array [
      Object {
        "units": "MINUTE",
        "value": "2",
      },
    ],
    "total": Array [
      Object {
        "units": "MINUTE",
        "value": "20",
      },
      Object {
        "units": "HOUR",
        "value": "1",
      },
    ],
  },
  "title": "my first recipe",
}
`;

exports[`getRecipeData valid id 1`] = `
Object {
  "recipe": Object {
    "createdAt": "2020-05-10 00:00:45.511 +00:00",
    "description": "the best recipe ever",
    "directionSections": Array [
      Object {
        "createdAt": "2020-05-10 00:07:45.511 +00:00",
        "directionSteps": Array [
          Object {
            "createdAt": "2020-05-10 00:09:45.511 +00:00",
            "directionSectionId": 2,
            "id": 1,
            "text": "step 1",
            "updatedAt": "2020-05-10 00:10:45.511 +00:00",
          },
          Object {
            "createdAt": "2020-05-10 00:11:45.511 +00:00",
            "directionSectionId": 2,
            "id": 2,
            "text": "step 2",
            "updatedAt": "2020-05-10 00:12:45.511 +00:00",
          },
        ],
        "id": 2,
        "label": "section 1",
        "recipeId": 123,
        "updatedAt": "2020-05-10 00:08:45.511 +00:00",
      },
    ],
    "id": 123,
    "ingredientSections": Array [
      Object {
        "createdAt": "2020-05-10 00:07:45.511 +00:00",
        "id": 1,
        "ingredients": Array [
          Object {
            "amount": "2",
            "createdAt": "2020-05-10 00:07:45.511 +00:00",
            "id": 1,
            "ingredientSectionId": 2,
            "name": "apples",
            "optional": true,
            "prep": "chopped",
            "toTaste": true,
            "unit": "cup",
            "updatedAt": "2020-05-10 00:08:45.511 +00:00",
          },
          Object {
            "createdAt": "2020-05-10 00:07:45.511 +00:00",
            "id": 2,
            "ingredientSectionId": 2,
            "name": "garlic",
            "optional": false,
            "prep": "minced",
            "rangedAmount": Object {
              "createdAt": "2020-05-10 00:07:45.511 +00:00",
              "id": 1,
              "ingredientId": 2,
              "max": "10",
              "min": "5",
              "updatedAt": "2020-05-10 00:08:45.511 +00:00",
            },
            "toTaste": true,
            "unit": "tablespoon",
            "updatedAt": "2020-05-10 00:08:45.511 +00:00",
          },
        ],
        "label": "ingredient section 1",
        "recipeId": 123,
        "updatedAt": "2020-05-10 00:08:45.511 +00:00",
      },
    ],
    "photo_url": "http://some.url.com/pic.jpg",
    "servings": 2,
    "source_display": "Mandy's Kitchen",
    "source_url": "http://some.url.com/recipe",
    "tags": Array [
      Object {
        "createdAt": "2020-05-10 00:07:45.511 +00:00",
        "id": 1,
        "label": "tag 1",
        "recipeId": 123,
        "recipeTag": Object {
          "createdAt": "2020-05-23T00:20:58.070Z",
          "recipeId": 123,
          "tagId": 12,
          "updatedAt": "2020-05-23T00:20:58.070Z",
        },
        "tagType": Object {
          "createdAt": "2020-05-22T22:25:13.475Z",
          "id": 1,
          "label": "category",
          "updatedAt": "2020-05-22T22:25:13.475Z",
        },
        "tagTypeId": 1,
        "tagid": 12,
        "updatedAt": "2020-05-10 00:08:45.511 +00:00",
      },
      Object {
        "createdAt": "2020-05-10 00:07:45.511 +00:00",
        "id": 2,
        "label": "tag 2",
        "recipeId": 123,
        "recipeTag": Object {
          "createdAt": "2020-05-23T00:20:58.070Z",
          "recipeId": 123,
          "tagId": 23,
          "updatedAt": "2020-05-23T00:20:58.070Z",
        },
        "tagType": Object {
          "createdAt": "2020-05-22T22:25:13.475Z",
          "id": 1,
          "label": "category",
          "updatedAt": "2020-05-22T22:25:13.475Z",
        },
        "tagTypeId": 1,
        "tagid": 23,
        "updatedAt": "2020-05-10 00:08:45.511 +00:00",
      },
    ],
    "timings": Array [
      Object {
        "createdAt": "2020-05-10 00:02:45.511 +00:00",
        "id": 1,
        "recipeId": 123,
        "type": "PREP",
        "units": "MINUTE",
        "updatedAt": "2020-05-10 00:03:45.511 +00:00",
        "value": "2",
      },
      Object {
        "createdAt": "2020-05-10 00:04:45.511 +00:00",
        "id": 2,
        "recipeId": 123,
        "type": "TOTAL",
        "units": "MINUTE",
        "updatedAt": "2020-05-10 00:05:45.511 +00:00",
        "value": "20",
      },
      Object {
        "createdAt": "2020-05-10 00:06:45.511 +00:00",
        "id": 3,
        "recipeId": 123,
        "type": "TOTAL",
        "units": "HOUR",
        "updatedAt": "2020-05-10 00:07:45.511 +00:00",
        "value": "1",
      },
    ],
    "title": "my first recipe",
    "updatedAt": "2020-05-10 00:01:45.511 +00:00",
  },
}
`;

exports[`updateRecipe id not found 1`] = `
Object {
  "message": "ID not found",
  "recipe": null,
  "success": false,
}
`;
